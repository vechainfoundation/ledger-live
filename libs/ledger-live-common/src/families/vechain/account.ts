import type { Account, Operation } from "@ledgerhq/types-live";
import { getAccountUnit } from "../../account";
import { formatCurrencyUnit } from "../../currencies";
import { Unit } from "@ledgerhq/types-cryptoassets";
import { getCryptoCurrencyById } from "../../currencies";

function formatAccountSpecifics(account: Account): string {
  const { energy } = account;
  if (!energy) {
    throw new Error("mycoin account expected");
  }

  const unit = getAccountUnit(account);
  const formatConfig = {
    disableRounding: true,
    alwaysShowSign: false,
    showCode: true,
  };

  let str = " ";

  str +=
    formatCurrencyUnit(unit, account.spendableBalance, formatConfig) +
    " spendable. ";

  if (energy.energy) {
    str += account.energy
      ? `${formatCurrencyUnit(
          getCryptoCurrencyById("vechainThor").units[0],
          account.energy.energy,
          formatConfig
        )}\n`
      : "no_VTHO_data\n";
    str += account.energy ? `\n\nOperations VTHO:\n` : "";
    if (account.energy?.transactions)
      account.energy.transactions.forEach((c) => {
        str +=
          (c.type == "IN" ? "+" : "-") +
          `${formatCurrencyUnit(
            getCryptoCurrencyById("vechainThor").units[0],
            c.value,
            formatConfig
          )}\t` +
          (c.type == "IN" ? "IN" : "OUT") +
          "\t" +
          c.recipients[0] +
          "\t" +
          c.date +
          "\n";
      });
  }

  return str;
}

function formatOperationSpecifics(op: Operation, unit: Unit): string {
  const { additionalField } = op.extra;

  let str = " ";

  const formatConfig = {
    disableRounding: true,
    alwaysShowSign: false,
    showCode: true,
  };

  str +=
    additionalField && !additionalField.isNaN()
      ? `\n    additionalField: ${
          unit
            ? formatCurrencyUnit(unit, additionalField, formatConfig)
            : additionalField
        }`
      : "";

  return str;
}

export default {
  formatAccountSpecifics,
  formatOperationSpecifics,
};
