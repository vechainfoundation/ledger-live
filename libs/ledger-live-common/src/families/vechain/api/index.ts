import { getAccount, getOperations } from "./sdk";

export default { getAccount, getOperations };
