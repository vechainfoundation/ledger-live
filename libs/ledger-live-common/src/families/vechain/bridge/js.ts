import type { Transaction, VechainAccount } from "../types";
import { makeAccountBridgeReceive } from "../../../bridge/jsHelpers";

import { sync, scanAccounts } from "../js-synchronisation";
import { AccountBridge, CurrencyBridge } from "@ledgerhq/types-live";
import {
  createTransaction,
  updateTransaction,
  prepareTransaction,
} from "../js-transaction";
import getTransactionStatus from "../js-getTransactionStatus";
import signOperation from "../js-signOperation";
import broadcast from "../js-broadcast";
import BigNumber from "bignumber.js";

const receive: AccountBridge<Transaction>["receive"] =
  makeAccountBridgeReceive();

const currencyBridge: CurrencyBridge = {
  scanAccounts,
  preload: async (): Promise<Record<string, any>> => {
    return {};
  },
  hydrate: (): void => {},
};

// TODO: Handle VTHO
const estimateMaxSpendable = async (inputs: {
  account: VechainAccount;
}): Promise<BigNumber> => {
  return inputs.account.balance;
};

const accountBridge: AccountBridge<Transaction> = {
  estimateMaxSpendable,
  createTransaction,
  updateTransaction,
  getTransactionStatus,
  prepareTransaction,
  sync,
  receive,
  signOperation,
  broadcast,
};

export default { currencyBridge, accountBridge };
