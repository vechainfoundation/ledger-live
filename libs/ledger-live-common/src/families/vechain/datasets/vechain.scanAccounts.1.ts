export default {
  name: "vechain seed 1",
  apdus: `
  => e002000015058000002c80000332800000000000000000000000
  <= 410482c44bc99cdf08d4360c97fbf62d387288ab75c576926943ad90059002720e93f58799391393c98ad41136aa4ac871b103d25cb9a88f1aadd7dbbe3c7794888928333234373631393364346133323438383332326666426239383335613763463263376532323032439000
  => e002000015058000002c80000332800000000000000000000001
  <= 410447267cc778e6f13242ae7f91b0be417746e59073d952a8c39900343f9de3e478f20221c3fff122ffefb30fe6b344e929561acd5d1060936fad036b1c57a1ccd128433039663137453965333542436130446537353339413433373246433337443934313733363731379000
  => e002000015058000002c80000332800000000000000000000002
  <= 4104e83afcbbec62e097093bcc24b74a259876fb86a0c5c39aa1ed9d850870c5f8ac05ad4e11d2bbfefbca2768518c5d57118ffcb67ad0c4a7d582ac94a618fb882828343438323531634436383163436363373931463234654130444432364233384239363633614537349000
  => e002000015058000002c80000332800000000000000000000003
  <= 41042351c71fa22ea9efdfbafbdd3366a7603fdd4ba52832fa4affebeb506dd40ba221ccd03c8b88f20a4d4c46c82edb9a98b85f3d8792b59baef0be1bb5a55335fa28303837466239306433374132453234363243436346454444373465434232346336383332393734319000
  => e002000015058000002c80000332800000000000000000000004
  <= 41049b56394e897da56b1999af6b63fba884becbc8750bad5a4cf7a0133e318ff0d08a08d78148fb7962e1328f18547830f012c2a07aef6fae281f0cf4f94ed0278228363063623139384144336546333830423734353034443536363134353635353441353039316135349000
  => e002000015058000002c80000332800000000000000000000005
  <= 4104001b56209f81d70da8ad502380512eab2273487a213928239604450472196e6a79b45e98743f83b8e7e29ca14cb21f2d6582db62d48ed013c87b423b745c4c5428303837613537386361323837394538443233343562614233423030393343664532416133643734419000
  => e002000015058000002c80000332800000000000000000000006
  <= 4104a50c733851da68ed7903acc1202dbfa147bd9ea2dfc02b23e0337885f90a81fb550b8c78b1305ebf49552dd6ae472356e107be49bbed3df8c48ffb6bb1a1519b28344239373138333832393944346530303934323635383039323035353144374332646242623963639000
  `,
};
