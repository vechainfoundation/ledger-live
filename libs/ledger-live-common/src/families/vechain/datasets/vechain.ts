import { AccountRaw } from "@ledgerhq/types-live";
export const vechain1: AccountRaw = {
  id: "js:2:vechain:0x087Fb90d37A2E2462CCcFEDD74eCB24c68329741:vechain",
  seedIdentifier:
    "0482c44bc99cdf08d4360c97fbf62d387288ab75c576926943ad90059002720e93f58799391393c98ad41136aa4ac871b103d25cb9a88f1aadd7dbbe3c77948889",
  name: "VeChain 4",
  starred: false,
  used: false,
  derivationMode: "vechain",
  index: 3,
  freshAddress: "0x087Fb90d37A2E2462CCcFEDD74eCB24c68329741",
  freshAddressPath: "44'/818'/0'/0/3",
  freshAddresses: [],
  blockHeight: 0,
  creationDate: new Date(1675317600).toISOString(),
  operationsCount: 1,
  operations: [
    {
      accountId:
        "js:2:vechain:0x32476193d4a32488322ffBb9835a7cF2c7e2202C:vechain+vechain%2Fvtho",
      blockHash:
        "0x00ddd8047ce47f277e3a021fee267bc025a68e9669e9e25fe0839a037515603a",
      blockHeight: 14538756,
      date: new Date().toISOString(),
      extra: {},
      hash: "0x05ee616beb80ddd8e41cec0bd12829f21395e99c13f7e2e2ea313f20f7414c2a",
      id: "0x05ee616beb80ddd8e41cec0bd12829f21395e99c13f7e2e2ea313f20f7414c2a",
      operator: undefined,
      recipients: ["0x32476193d4a32488322ffbb9835a7cf2c7e2202c"],
      senders: ["0x4f6fc409e152d33843cf4982d414c1dd0879277e"],
      standard: undefined,
      tokenId: undefined,
      type: "IN",
      value: "10000000000000000000",
      fee: "2100000000000000",
    },
  ],
  pendingOperations: [],
  currencyId: "vechain",
  unitMagnitude: 18,
  lastSyncDate: "2023-02-28T09:07:53.785Z",
  balance: "10000000000000000000",
  spendableBalance: "10000000000000000000",
  balanceHistoryCache: {
    HOUR: { balances: [], latestDate: 1677574800000 },
    DAY: { balances: [], latestDate: 1677538800000 },
    WEEK: { balances: [], latestDate: 1677366000000 },
  },
  subAccounts: [
    {
      type: "TokenAccountRaw",
      id: "js:2:vechain:0x087Fb90d37A2E2462CCcFEDD74eCB24c68329741:vechain+vechain%2Fvtho",
      parentId:
        "js:2:vechain:0x087Fb90d37A2E2462CCcFEDD74eCB24c68329741:vechain",
      starred: false,
      tokenId: "vechain/vtho",
      balance: "10000000000000000000",
      spendableBalance: "9000000000000000000",
      balanceHistoryCache: {
        HOUR: { latestDate: null, balances: [] },
        DAY: { latestDate: null, balances: [] },
        WEEK: { latestDate: null, balances: [] },
      },
      creationDate: "2023-02-28T09:07:53.785Z",
      operationsCount: 0,
      operations: [],
      pendingOperations: [],
      swapHistory: [],
    },
  ],
  swapHistory: [],
};
