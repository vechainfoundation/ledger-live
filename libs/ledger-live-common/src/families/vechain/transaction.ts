import type {
  Transaction,
  TransactionRaw,
  TransactionStatus,
  TransactionStatusRaw,
} from "./types";
import {
  formatTransactionStatusCommon as formatTransactionStatus,
  fromTransactionCommonRaw,
  fromTransactionStatusRawCommon,
  toTransactionCommonRaw,
  toTransactionStatusRawCommon,
} from "@ledgerhq/coin-framework/transaction/common";

export const formatTransaction = (t: Transaction): string => {
  return "";
};

export const fromTransactionRaw = (tr: TransactionRaw): Transaction => {
  const common = fromTransactionCommonRaw(tr);

  return {
    ...tr,
    ...common,
  };
};

export const toTransactionRaw = (t: Transaction): TransactionRaw => {
  const common = toTransactionCommonRaw(t);

  return {
    ...t,
    ...common,
  };
};

export const fromTransactionStatusRaw = (
  ts: TransactionStatusRaw
): TransactionStatus => {
  const common = fromTransactionStatusRawCommon(ts);

  return {
    ...ts,
    ...common,
  };
};

export const toTransactionStatusRaw = (
  ts: TransactionStatus
): TransactionStatusRaw => {
  const common = toTransactionStatusRawCommon(ts);

  return {
    ...ts,
    ...common,
  };
};

export default {
  formatTransaction,
  formatTransactionStatus,
  fromTransactionRaw,
  toTransactionRaw,
  fromTransactionStatusRaw,
  toTransactionStatusRaw,
};
