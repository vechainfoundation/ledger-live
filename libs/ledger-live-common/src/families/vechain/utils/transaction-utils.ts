import { DEFAULT_GAS_COEFFICIENT, HEX_PREFIX } from "../constants";
import crypto from "crypto";
import BigNumber from "bignumber.js";
import { Transaction as ThorTransaction } from "thor-devkit";
import params from "../contracts/abis/params";
import { BASE_GAS_PRICE_KEY, PARAMS_ADDRESS } from "../contracts/constants";
import { Query } from "../api/types";
import { query } from "../api/sdk";
import { Account, TokenAccount } from "@ledgerhq/types-live";
import { Transaction, TransactionInfo } from "../types";
import { isValid } from "./address-utils";
import { calculateClausesVtho } from "../js-transaction";

const GAS_COEFFICIENT = 15000;

/**
 * Generate a Unique ID to be used as a nonce
 * @returns a unique string
 */
export const generateNonce = (): string => {
  const randBuffer = crypto.randomBytes(Math.ceil(4));
  if (!randBuffer) throw Error("Failed to generate random hex");
  return `${HEX_PREFIX}${randBuffer.toString("hex").substring(0, 8)}`;
};

/**
 * Estimate the gas that will be used by the transaction.
 * @param transaction - The transaction to estimate the gas for
 * @returns an estimate of the gas usage
 */
export const estimateGas = async (t: Transaction): Promise<number> => {
  const intrinsicGas = ThorTransaction.intrinsicGas(t.body.clauses);

  const queryData: Query[] = [];

  t.body.clauses.forEach(c => {
    if (c.to) {
      queryData.push({
        to: c.to,
        data: c.data,
      });
    }
  });

  const response = await query(queryData);

  const execGas = response.reduce((sum, out) => sum + out.gasUsed, 0);

  return intrinsicGas + (execGas ? execGas + GAS_COEFFICIENT : 0);
};

const getBaseGasPrice = async (): Promise<string> => {
  const queryData: Query = {
    to: PARAMS_ADDRESS,
    data: params.get.encode(BASE_GAS_PRICE_KEY),
  };

  const response = await query([queryData]);

  // Expect 1 value
  if (response && response.length != 1) throw Error("Unexpected response received for query");

  return response[0].data;
};

/**
 * Calculate the fee in VTHO
 * @param gas - the gas used
 * @param gasPriceCoef - the gas price coefficient
 * @returns the fee in VTHO
 */
export const calculateFee = async (gas: BigNumber, gasPriceCoef: number): Promise<BigNumber> => {
  const baseGasPrice = await getBaseGasPrice();
  return new BigNumber(baseGasPrice).times(gasPriceCoef).idiv(255).plus(baseGasPrice).times(gas);
};

export const calculateTransactionInfo = async (
  account: Account,
  transaction: Transaction,
  fixedMaxTokenFees?: BigNumber,
): Promise<TransactionInfo> => {
  const { subAccounts } = account;
  const { amount: oldAmount, useAllAmount, subAccountId } = transaction;
  const maxTokenFees = fixedMaxTokenFees || (await calculateMaxFeesToken(transaction));

  const tokenAccount =
    subAccountId && subAccounts
      ? (subAccounts.find(subAccount => {
          return subAccount.id === subAccountId;
        }) as TokenAccount)
      : undefined;
  const isTokenAccount = !!tokenAccount;

  let balance;
  let spendableBalance;
  let amount;

  if (isTokenAccount) {
    balance = tokenAccount.balance;
    spendableBalance = tokenAccount.balance.minus(maxTokenFees).gt(0)
      ? tokenAccount.balance.minus(maxTokenFees)
      : new BigNumber(0);
    amount = useAllAmount ? spendableBalance : oldAmount;
  } else {
    balance = account.balance;
    spendableBalance = account.balance;
    amount = useAllAmount ? spendableBalance : oldAmount;
  }

  return {
    isTokenAccount,
    amount,
    spendableBalance,
    balance,
    tokenAccount,
  };
};

export const calculateMaxFeesToken = async (transaction: Transaction): Promise<BigNumber> => {
  if (transaction.recipient && isValid(transaction.recipient)) {
    const clauses = await calculateClausesVtho(transaction, transaction.amount);
    const gas = await estimateGas({
      ...transaction,
      body: { ...transaction.body, clauses: clauses },
    });
    return await calculateFee(new BigNumber(gas), DEFAULT_GAS_COEFFICIENT);
  }
  return new BigNumber(0);
};

export const calculateTotalSpent = (isToken: boolean, transaction: Transaction): BigNumber => {
  if (isToken) {
    return transaction.amount.plus(transaction.estimatedFees);
  } else {
    return transaction.amount;
  }
};
