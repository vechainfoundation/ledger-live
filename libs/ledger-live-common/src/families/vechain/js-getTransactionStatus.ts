import {
  AmountRequired,
  FeeNotLoaded,
  InvalidAddress,
  InvalidAddressBecauseDestinationIsAlsoSource,
  NotEnoughBalance,
  RecipientRequired,
} from "@ledgerhq/errors";
import type { TransactionStatus } from "./types";
import type { Transaction } from "./types";
import { Account } from "@ledgerhq/types-live";
import { calculateTransactionInfo } from "./utils/calculateTransactionInfo";
import { isValid } from "./utils/address-utils";

// NOTE: seems like the spendableBalance is not updated correctly:
// use balance.minus(estimatedFees) instead
const getTransactionStatus = async (
  account: Account,
  transaction: Transaction
): Promise<TransactionStatus> => {
  const { freshAddress, currency, subAccounts } = account;
  const { body, recipient } = transaction;
  const errors: Record<string, Error> = {};
  const warnings: Record<string, Error> = {};
  const {
    amount,
    isTokenAccount,
    estimatedFees,
    totalSpent,
    spendableBalance,
  } = await calculateTransactionInfo(account, transaction);

  // TODO: Implement this properly
  if (!body || !body.gas) {
    errors["body"] = new FeeNotLoaded();
  }

  if (!recipient) {
    errors.recipient = new RecipientRequired();
  } else if (freshAddress === recipient) {
    warnings.recipient = new InvalidAddressBecauseDestinationIsAlsoSource();
  } else if (!isValid(recipient)) {
    errors.recipient = new InvalidAddress("", {
      currencyName: currency.name,
    });
  }
  // TODO: add a validation function

  if (!amount.gt(0)) {
    errors.amount = new AmountRequired();
  } else {
    if (amount.gt(spendableBalance)) {
      errors.amount = new NotEnoughBalance();
    }
    if (!isTokenAccount) {
      // vet
      const vthoBalance = subAccounts?.[0].balance;
      if (estimatedFees.gt(vthoBalance || 0)) {
        errors.amount = new NotEnoughBalance();
      }
    }
  }

  return Promise.resolve({
    errors,
    warnings,
    estimatedFees,
    amount,
    totalSpent,
  });
};

export default getTransactionStatus;
