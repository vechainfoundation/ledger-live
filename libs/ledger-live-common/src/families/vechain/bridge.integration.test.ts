import "../../__tests__/test-helpers/setup";
import { testBridge } from "../../__tests__/test-helpers/bridge";
import type { AccountRaw, DatasetTest } from "@ledgerhq/types-live";
import type { Transaction } from "./types";
import { fromTransactionRaw } from "./transaction";
import { DEFAULT_GAS_COEFFICIENT, TESTNET_CHAIN_TAG } from "./constants";
import { vechain1 } from "./datasets/vechain";
import { generateNonce } from "./utils/transaction-utils";
import BigNumber from "bignumber.js";
import {
  setSupportedCurrencies,
  listSupportedCurrencies,
  getCryptoCurrencyById,
} from "../../currencies";
import vechainScanAccounts1 from "./datasets/vechain.scanAccounts.1";
import { AmountRequired, NotEnoughBalance } from "@ledgerhq/errors";
import VIP180 from "./contracts/abis/VIP180";
import { CryptoCurrencyId } from "@ledgerhq/types-cryptoassets";

const listSupported = listSupportedCurrencies();
listSupported.push(getCryptoCurrencyById("vechain"));
setSupportedCurrencies(listSupported.map((c) => c.id) as CryptoCurrencyId[]);

const dataset: DatasetTest<Transaction> = {
  implementations: ["js"],
  currencies: {
    vechain: {
      FIXME_ignoreAccountFields: [
        "balance",
        "spendableBalance",
        "estimateMaxSpendable",
        "creationDate",
        "blockRef",
      ],
      FIXME_ignoreOperationFields: ["nonce"],
      scanAccounts: [vechainScanAccounts1],
      accounts: [
        {
          implementations: ["mock", "js"],
          transactions: [
            {
              name: "Send VET",
              transaction: fromTransactionRaw({
                family: "vechain",
                mode: "send_vet",
                recipient: "0x17733CAb76d9A2112576443F21735789733B1ca3",
                amount: "10000000000000000000",
                body: {
                  chainTag: TESTNET_CHAIN_TAG,
                  blockRef: "0x00634a0c856ec1db",
                  expiration: 18,
                  clauses: [
                    {
                      to: "0x17733CAb76d9A2112576443F21735789733B1ca3",
                      value: "10000000000000000000",
                      data: "0x",
                    },
                  ],
                  gasPriceCoef: DEFAULT_GAS_COEFFICIENT,
                  gas: "0",
                  dependsOn: null,
                  nonce: generateNonce(),
                },
              }),
              expectedStatus: {
                amount: new BigNumber("10000000000000000000"),
                estimatedFees: new BigNumber("210000000000000000"),
                totalSpent: new BigNumber("10000000000000000000"),
                errors: {},
                warnings: {},
              },
            },
            {
              name: "Send VTHO",
              transaction: fromTransactionRaw({
                family: "vechain",
                mode: "send_vtho",
                recipient: "0x17733CAb76d9A2112576443F21735789733B1ca3",
                amount: "9000000000000000000",
                body: {
                  chainTag: TESTNET_CHAIN_TAG,
                  blockRef: "0x00634a0c856ec1db",
                  expiration: 18,
                  clauses: [
                    {
                      to: "0x0000000000000000000000000000456e65726779",
                      value: 0,
                      data: VIP180.transfer.encode(
                        "0x17733CAb76d9A2112576443F21735789733B1ca3",
                        "9000000000000000000"
                      ),
                    },
                  ],
                  gasPriceCoef: DEFAULT_GAS_COEFFICIENT,
                  gas: "0",
                  dependsOn: null,
                  nonce: generateNonce(),
                },
              }),
              expectedStatus: {
                amount: new BigNumber("9000000000000000000"),
                estimatedFees: new BigNumber("385030000000000000"),
                totalSpent: new BigNumber("9000000000000000000"),
                errors: {},
                warnings: {},
              },
            },
            {
              name: "Amount required",
              transaction: fromTransactionRaw({
                family: "vechain",
                mode: "send_vet",
                recipient: "0x17733CAb76d9A2112576443F21735789733B1ca3",
                amount: "",
                body: {
                  chainTag: TESTNET_CHAIN_TAG,
                  blockRef: "0x00634a0c856ec1db",
                  expiration: 18,
                  clauses: [{ to: "", value: 0, data: "0x" }],
                  gasPriceCoef: DEFAULT_GAS_COEFFICIENT,
                  gas: "0",
                  dependsOn: null,
                  nonce: generateNonce(),
                },
              }),
              expectedStatus: {
                errors: {
                  amount: new AmountRequired(),
                },
                warnings: {},
              },
            },
            {
              name: "VET balance not enough",
              transaction: fromTransactionRaw({
                family: "vechain",
                mode: "send_vet",
                recipient: "0x17733CAb76d9A2112576443F21735789733B1ca3",
                amount: "2000000000000000000000",
                body: {
                  chainTag: TESTNET_CHAIN_TAG,
                  blockRef: "0x00634a0c856ec1db",
                  expiration: 18,
                  clauses: [
                    {
                      to: "0x17733CAb76d9A2112576443F21735789733B1ca3",
                      value: "2000000000000000000000",
                      data: "0x",
                    },
                  ],
                  gasPriceCoef: DEFAULT_GAS_COEFFICIENT,
                  gas: "0",
                  dependsOn: null,
                  nonce: generateNonce(),
                },
              }),
              expectedStatus: {
                amount: new BigNumber("2000000000000000000000"),
                errors: {
                  amount: new NotEnoughBalance(),
                },
                warnings: {},
                totalSpent: new BigNumber("0"),
                estimatedFees: new BigNumber("0"),
              },
            },
            {
              name: "VTHO balance not enough",
              transaction: fromTransactionRaw({
                family: "vechain",
                mode: "send_vtho",
                recipient: "0x17733CAb76d9A2112576443F21735789733B1ca3",
                amount: "2000000000000000000000",
                body: {
                  chainTag: TESTNET_CHAIN_TAG,
                  blockRef: "0x00634a0c856ec1db",
                  expiration: 18,
                  clauses: [
                    {
                      to: "0x17733CAb76d9A2112576443F21735789733B1ca3",
                      value: "2000000000000000000000",
                      data: "0x",
                    },
                  ],
                  gasPriceCoef: DEFAULT_GAS_COEFFICIENT,
                  gas: "0",
                  dependsOn: null,
                  nonce: generateNonce(),
                },
              }),
              expectedStatus: {
                amount: new BigNumber("2000000000000000000000"),
                errors: {
                  amount: new NotEnoughBalance(),
                },
                warnings: {},
                totalSpent: new BigNumber("0"),
                estimatedFees: new BigNumber("0"),
              },
            },
          ],
          raw: vechain1 as AccountRaw,
          FIXME_tests: [
            "balance is sum of ops",
            "empty transaction is equals to itself",
          ],
        },
      ],
    },
  },
};

testBridge(dataset);
