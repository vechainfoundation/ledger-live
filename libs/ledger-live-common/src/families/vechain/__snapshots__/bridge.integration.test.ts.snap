// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vechain currency bridge scanAccounts vechain seed 1 1`] = `
Array [
  Object {
    "balance": "0",
    "currencyId": "vechain",
    "derivationMode": "vechain",
    "feesCurrencyId": "vechain/vtho",
    "freshAddress": "0x82d984c42fC9d49E2d4CFFdcad59301AfFca7E02",
    "freshAddressPath": "44'/818'/0'/0/0",
    "freshAddresses": Array [
      Object {
        "address": "0x82d984c42fC9d49E2d4CFFdcad59301AfFca7E02",
        "derivationPath": "44'/818'/0'/0/0",
      },
    ],
    "id": "js:2:vechain:0x82d984c42fC9d49E2d4CFFdcad59301AfFca7E02:vechain",
    "index": 0,
    "name": "Vechain 1",
    "nfts": undefined,
    "operationsCount": 3,
    "pendingOperations": Array [],
    "seedIdentifier": "0428594655d5b5c95bb2c284ae480caae57df83d5d51e53f3cad0de76c014fdd9c14d47fc40abc126bf41197021d3a2ea6895ce15e50bdf255769b65cd87aaaa7f",
    "spendableBalance": "0",
    "starred": false,
    "subAccounts": Array [],
    "swapHistory": Array [],
    "syncHash": undefined,
    "unitMagnitude": 18,
    "used": true,
  },
  Object {
    "approvals": undefined,
    "balance": "2913372510904685500",
    "id": "js:2:vechain:0x82d984c42fC9d49E2d4CFFdcad59301AfFca7E02:vechain+vechain%2Fvtho",
    "operationsCount": 1,
    "parentId": "js:2:vechain:0x82d984c42fC9d49E2d4CFFdcad59301AfFca7E02:vechain",
    "pendingOperations": Array [],
    "spendableBalance": "2913372510904685500",
    "starred": false,
    "swapHistory": Array [],
    "tokenId": "vechain/vtho",
    "type": "TokenAccountRaw",
  },
  Object {
    "balance": "0",
    "currencyId": "vechain",
    "derivationMode": "vechain",
    "feesCurrencyId": "vechain/vtho",
    "freshAddress": "0x91425B8c5EBB822c1f60270DA3B2cB9337F8fAB8",
    "freshAddressPath": "44'/818'/0'/0/1",
    "freshAddresses": Array [
      Object {
        "address": "0x91425B8c5EBB822c1f60270DA3B2cB9337F8fAB8",
        "derivationPath": "44'/818'/0'/0/1",
      },
    ],
    "id": "js:2:vechain:0x91425B8c5EBB822c1f60270DA3B2cB9337F8fAB8:vechain",
    "index": 1,
    "name": "Vechain 2",
    "nfts": undefined,
    "operationsCount": 0,
    "pendingOperations": Array [],
    "seedIdentifier": "0428594655d5b5c95bb2c284ae480caae57df83d5d51e53f3cad0de76c014fdd9c14d47fc40abc126bf41197021d3a2ea6895ce15e50bdf255769b65cd87aaaa7f",
    "spendableBalance": "0",
    "starred": false,
    "subAccounts": Array [],
    "swapHistory": Array [],
    "syncHash": undefined,
    "unitMagnitude": 18,
    "used": false,
  },
  Object {
    "approvals": undefined,
    "balance": "0",
    "id": "js:2:vechain:0x91425B8c5EBB822c1f60270DA3B2cB9337F8fAB8:vechain+vechain%2Fvtho",
    "operationsCount": 0,
    "parentId": "js:2:vechain:0x91425B8c5EBB822c1f60270DA3B2cB9337F8fAB8:vechain",
    "pendingOperations": Array [],
    "spendableBalance": "0",
    "starred": false,
    "swapHistory": Array [],
    "tokenId": "vechain/vtho",
    "type": "TokenAccountRaw",
  },
]
`;

exports[`vechain currency bridge scanAccounts vechain seed 1 2`] = `
Array [
  Array [
    Object {
      "accountId": "js:2:vechain:0x82d984c42fC9d49E2d4CFFdcad59301AfFca7E02:vechain",
      "blockHash": "0x00f3c1b82e4a5668e0ce880387d0ee5637ea9b8b132f652f96d627fb54ff3548",
      "blockHeight": 15974840,
      "contract": undefined,
      "extra": Object {},
      "fee": "210000000000000000",
      "hash": "0x612174c3c426978929e2a50bf8cfdfc48570cc1816ca171f23622d6a42d694bd",
      "id": "js:2:vechain:0x82d984c42fC9d49E2d4CFFdcad59301AfFca7E02:vechain-0x612174c3c426978929e2a50bf8cfdfc48570cc1816ca171f23622d6a42d694bd-IN",
      "operator": undefined,
      "recipients": Array [
        "0x82d984c42fc9d49e2d4cffdcad59301affca7e02",
      ],
      "senders": Array [
        "0xa0ae9a1c595753c001cf56fec9f730bf32b8f592",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "type": "IN",
      "value": "1526316870530000000000",
    },
    Object {
      "accountId": "js:2:vechain:0x82d984c42fC9d49E2d4CFFdcad59301AfFca7E02:vechain",
      "blockHash": "0x00e742f27e0dddaa6df6fc23fb39a24acb50d57dbbe114f64fb06ddfa4d358e8",
      "blockHeight": 15155954,
      "contract": undefined,
      "extra": Object {},
      "fee": "210000000000000000",
      "hash": "0x901109642a6f56495bdd652cd7c9fb96ee476377b976551b46a05437567af58c",
      "id": "js:2:vechain:0x82d984c42fC9d49E2d4CFFdcad59301AfFca7E02:vechain-0x901109642a6f56495bdd652cd7c9fb96ee476377b976551b46a05437567af58c-IN",
      "operator": undefined,
      "recipients": Array [
        "0x82d984c42fc9d49e2d4cffdcad59301affca7e02",
      ],
      "senders": Array [
        "0x79028e3d948bd5873ccf58a69089cac105832129",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "type": "IN",
      "value": "3000000000000000000",
    },
    Object {
      "accountId": "js:2:vechain:0x82d984c42fC9d49E2d4CFFdcad59301AfFca7E02:vechain",
      "blockHash": "0x00f3c1bf144baddcd7b58a3c888a843f1d9a7a0f34f5779e22d23e601ebba8bb",
      "blockHeight": 15974847,
      "contract": undefined,
      "extra": Object {},
      "fee": "210000000000000000",
      "hash": "0xd46e52925482e16c0d30437f347fcb511849b5e3fc1aa5ab8f13c487c1e59142",
      "id": "js:2:vechain:0x82d984c42fC9d49E2d4CFFdcad59301AfFca7E02:vechain-0xd46e52925482e16c0d30437f347fcb511849b5e3fc1aa5ab8f13c487c1e59142-OUT",
      "operator": undefined,
      "recipients": Array [
        "0x79028e3d948bd5873ccf58a69089cac105832129",
      ],
      "senders": Array [
        "0x82d984c42fc9d49e2d4cffdcad59301affca7e02",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "type": "OUT",
      "value": "1529316870530000000000",
    },
  ],
  Array [
    Object {
      "accountId": "js:2:vechain:0x82d984c42fC9d49E2d4CFFdcad59301AfFca7E02:vechain+vechain%2Fvtho",
      "blockHash": "0x00e742f48ff7553a1b5276886ca7df27444cafa2086757393ce31ff68422a4b9",
      "blockHeight": 15155956,
      "contract": undefined,
      "extra": Object {},
      "fee": "365180000000000000",
      "hash": "0xa3049ffd1af19c85659e76e50e733098df194ba9df7597e6c654d688c553488b",
      "id": "js:2:vechain:0x82d984c42fC9d49E2d4CFFdcad59301AfFca7E02:vechain+vechain%2Fvtho-0xa3049ffd1af19c85659e76e50e733098df194ba9df7597e6c654d688c553488b-IN",
      "operator": undefined,
      "recipients": Array [
        "0x82d984c42fc9d49e2d4cffdcad59301affca7e02",
      ],
      "senders": Array [
        "0x79028e3d948bd5873ccf58a69089cac105832129",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "type": "IN",
      "value": "3000000000000000000",
    },
  ],
  Array [],
  Array [],
]
`;
